File: Day080_2025/Q080_Leetcode.py
from collections import deque, defaultdict
from typing import List

class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        available_supplies = set(supplies)
        recipe_to_index = {recipe: i for i, recipe in enumerate(recipes)}
        dependency_graph = defaultdict(list)
        in_degree = [0] * len(recipes)

        for i, ingredient_list in enumerate(ingredients):
            for ingredient in ingredient_list:
                if ingredient not in available_supplies:
                    dependency_graph[ingredient].append(recipes[i])
                    in_degree[i] += 1

        queue = deque([i for i in range(len(recipes)) if in_degree[i] == 0])
        created_recipes = []

        while queue:
            recipe_idx = queue.popleft()
            recipe = recipes[recipe_idx]
            created_recipes.append(recipe)

            for dependent_recipe in dependency_graph[recipe]:
                dependent_idx = recipe_to_index[dependent_recipe]
                in_degree[dependent_idx] -= 1
                if in_degree[dependent_idx] == 0:
                    queue.append(dependent_idx)

        return created_recipes

File: Day080_2025/Q080_Leetcode.java
import java.util.*;

class Q080_Leetcode {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        Set<String> availableSupplies = new HashSet<>(Arrays.asList(supplies));
        Map<String, Integer> recipeToIndex = new HashMap<>();
        Map<String, List<String>> dependencyGraph = new HashMap<>();
        int n = recipes.length;
        
        for (int i = 0; i < n; i++) {
            recipeToIndex.put(recipes[i], i);
        }
        
        int[] inDegree = new int[n];
        
        for (int i = 0; i < n; i++) {
            for (String ingredient : ingredients.get(i)) {
                if (!availableSupplies.contains(ingredient)) {
                    dependencyGraph.computeIfAbsent(ingredient, _ -> new ArrayList<>()).add(recipes[i]);
                    inDegree[i]++;
                }
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        List<String> createdRecipes = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }
        
        while (!queue.isEmpty()) {
            int recipeIdx = queue.poll();
            String recipe = recipes[recipeIdx];
            createdRecipes.add(recipe);
            
            if (!dependencyGraph.containsKey(recipe)) continue;
            
            for (String dependentRecipe : dependencyGraph.get(recipe)) {
                int dependentIdx = recipeToIndex.get(dependentRecipe);
                if (--inDegree[dependentIdx] == 0) {
                    queue.offer(dependentIdx);
                }
            }
        }
        
        return createdRecipes;
    }
}
